trinity/sync/full/state.py

```diff
@@ -245,11 +221,42 @@ def deregister_peer(self, peer: BasePeer) -> None:
                 return
 
             candidates = list(not_yet_requested.difference(self._peer_missing_nodes[peer]))
-            batch = candidates[:eth_constants.MAX_STATE_FETCH]
+            batch = tuple(candidates[:eth_constants.MAX_STATE_FETCH])
             not_yet_requested = not_yet_requested.difference(batch)
             self.request_tracker.active_requests[peer] = (time.time(), batch)
-            self.logger.debug("Requesting %d trie nodes to %s", len(batch), peer)
-            peer.sub_proto.send_get_node_data(batch)
+            asyncio.ensure_future(self._request_and_process_nodes(peer, batch))
+
+    async def _request_and_process_nodes(self, peer: ETHPeer, batch: Tuple[Hash32, ...]) -> None:
+        self.logger.debug("Requesting %d trie nodes from %s", len(batch), peer)
+        node_data = await peer.requests.get_node_data(batch)
+        self.request_tracker.active_requests.pop(peer)
+
+        self.logger.debug("Got %d NodeData entries from %s", len(node_data), peer)
+
+        if node_data:
+            node_keys, _ = zip(*node_data)
+        else:
+            node_keys = tuple()
+
+        # check for missing nodes and re-schedule them
+        missing = set(batch).difference(node_keys)
+
+        # TODO: this doesn't actually mean the peer doesn't have them, just
+        # that they didn't respond with them this time.  We should explore
+        # alternate ways to do this since a false negative here will result in
+        # not requesting this node from this peer again.
```

Comment:
Maybe if we have marked the node as missing from *all* connected peers, then we drop the missing node from all peers and try again?

Reply:
That could at least be a decent short term solution to the larger problem which is that we need to have state syncing continually update to newer state roots, otherwise it's basically impossible to finish because the nodes you are connected to are pruning old state history and in my experiments, you always end up with a few missing trie nodes, even if you have 99% of the state db already downloaded because of the time it takes to process the state trie and find the few nodes you are missing.